### PR TITLE
Add support for local connect terminal context

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for having a terminal-based context for the LightningCLI. This works by adding the paren pid in the connection folder.

Furthermore, every time a command runs, it will clean out the dead connections.

Fixes #\<issue_number>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/cli/commands/connection.py [hunk 1]
```diff
@@ -14,20 +15,24 @@
 from lightning_app.utilities.log import get_logfile
 from lightning_app.utilities.network import LightningClient
 
+_HOME = os.path.expanduser("~")
+_PPID = str(psutil.Process(os.getpid()).ppid())
```
Comment (awaelchli) [inline src/lightning_app/cli/commands/connection.py:19]: I don't think we install psutil by default. Might need to be added to base requirement if that's the only way to get the parent pid.

File: src/lightning_app/cli/commands/connection.py [hunk 2]
```diff
@@ -14,20 +15,24 @@
...
+_LIGHTNING_CONNECTION = os.path.join(_HOME, ".lightning", "lightning_connection")
+_LIGHTNING_FOLDER = os.path.join(_LIGHTNING_CONNECTION, _PPID)
```
Comment (awaelchli) [inline src/lightning_app/cli/commands/connection.py:21]: Maybe should have a different name. when it says "lightning folder', I would think it is the .lightning

**Review by awaelchli — APPROVED** [review]
Solid UX improvement!

File: src/lightning_app/cli/commands/connection.py (hunk 2 above)
Comment (tchaton) [reply src/lightning_app/cli/commands/connection.py:21]: Renamed to `_LIGHTNING_CONNECTION_FOLDER`

File: src/lightning_app/cli/commands/connection.py (hunk 1 above)
Comment (tchaton) [reply src/lightning_app/cli/commands/connection.py:19]: Yes, let's keep psutil. This is always useful to clean up process stuffs.

**Comment by Borda** [comment]
The CI shall be fixed with #15259

**Review by manskx — APPROVED** [review]

File: .azure/app-cloud-e2e.yml [hunk 3]
```diff
@@ -106,7 +106,7 @@ jobs:
         PACKAGE_NAME: app
       displayName: 'Install lightning app'
 
-    - bash: python -m pip install -r requirements/app/test.txt
+    - bash: python -m pip install -r requirements/app/test.txt  --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
```
Comment (carmocca) [inline .azure/app-cloud-e2e.yml:109]: nit: 2 spaces
```suggestion
    - bash: python -m pip install -r requirements/app/test.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
```

File: requirements/app/base.txt [hunk 4]
```diff
@@ -9,3 +9,4 @@ traitlets>=5.3.0, <=5.4.0
 arrow>=1.2.0, <1.2.4
 lightning-utilities==0.3.*
 beautifulsoup4<=4.8.2
+psutil
```
Comment (carmocca) [inline requirements/app/base.txt:12]: Might want to set a min/max version as in all other dependencies

**Review by carmocca — APPROVED** [review]

File: requirements/app/base.txt (hunk 4 above)
Comment (tchaton) [reply requirements/app/base.txt]: Done !

**Pull request merged**